db/commitlog/commitlog.cc

```diff
@@ -1626,21 +1626,19 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
         // the old one has terminated with either result or exception.
         // Do all waiting through the shared_future
         if (!_segment_allocating) {
-            _segment_allocating.emplace(new_segment().discard_result());
-        }
-        try {
-            co_await _segment_allocating->get_future(timeout);
-            // once we've managed to get a result, any of us, the
-            // shared_future should be released.
-            _segment_allocating = std::nullopt;
-        } catch (timed_out_error&) {
-            throw; // not thrown by new_segment. Just no result yet.
-        } catch (...) {            
-            // once we've managed to get a result, any of us, the
-            // shared_future should be released.
-            _segment_allocating = std::nullopt;
-            throw;
+            auto f = new_segment();
+            // must check that we are not already done.
+            if (f.available()) {
+                f.get(); // maybe force exception
+                continue;
+            }
+            _segment_allocating.emplace(f.discard_result().finally([this] {
```

Comment:
Is it ok to call discard_result() after the get()?
If, above, you wanted to run f.get() just for failure, you should check f.failure(), not f.available().

Reply:
> 1. I think the commit id referenced in the commit message is the wrong one (?), please check again.
> 
Yeah, copy-paste error in the message. Note the missing words too... Fixed

Reply:
The f.get() is followed by a continue. We jump back to loop edge. The `discard_result()` is only for the case where f was not available, and we did not call get.